pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/NopaMetric.java

```diff
@@ -0,0 +1,36 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.metrics.impl;
+
+import net.sourceforge.pmd.lang.java.ast.ASTAnyTypeDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTAnyTypeDeclaration.TypeKind;
+import net.sourceforge.pmd.lang.java.metrics.signature.JavaFieldSigMask;
+import net.sourceforge.pmd.lang.java.metrics.signature.JavaSignature.Visibility;
+import net.sourceforge.pmd.lang.metrics.MetricOptions;
+
+/**
+ * Number of public attributes.
+ *
+ * @author Clément Fournier
+ * @since 6.0.0
+ */
+public class NopaMetric extends AbstractJavaClassMetric {
```

Comment:
Interesting, the book has a small error. But "NOPA" for Number of public attributes makes definitely more sense than the wrong "NOAP" :)